core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -80,114 +146,344 @@
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@code start} a second time if
+    * {@link #stop()} has been invoked and allowed for its stage to complete.
     * @param ctx initialization context used to initialize this store
     * @return a stage that when complete signals that this store has been successfully started
     */
    CompletionStage<Void> start(InitializationContext ctx);
 
    /**
+    * This method will be invoked is being shut down. It is expected that all resources related to the store to be
+    * freed upon completion of the returned stage.
     * <p>
     * This method is guaranteed to not be invoked concurrently with other operations. This also means another method
     * will not be invoked on this store until after the returned Stage completes.
+    * <p>
+    * It is expected that an implementation should be able to "restart" by invoking {@link #start(InitializationContext)}
+    * a second time if {@code stop} has been invoked and allowed for its stage to complete.
     * @return a stage that when complete signals that this store has been stopped
     */
    CompletionStage<Void> stop();
 
    /**
-    *
-    * @return
+    * Returns a set of characteristics for this store and its elements. This method will be invoked one to many times
+    * to determine which methods of the store can be used and how its data can be handled.
+    * <p>
+    * Please see {@link Characteristic} and its values for a description of what each characteristic declares the
+    * store as supporting.
+    * @implSpec
+    * The default implementation returns an empty set
+    * @return the set of characteristics that this store supports
     */
    default Set<Characteristic> characteristics() {
       return EnumSet.noneOf(Characteristic.class);
    }
 
    /**
-    *
-    * @return
+    * Returns a stage that when complete will tell if the current store can be accessed for requests. This can be useful
+    * for some store implementations that may rely on an external source that may become unreachable. This can reduce
+    * sending requests to a store that is not available as future Cache requests will all be met with
+    * {@link StoreUnavailableException}.
+    * <p>
+    * Store availability is polled automatically every so often to update the status if the underlying store changes
+    * availability. This method will not be invoked concurrently with itself (ie. this method will not be invoked
+    * until after the previous stage has completed), but will be invoked concurrently with other operations, excluding
+    * start and stop.
+    * <p>
+    * The other benefit of implementing this method is if the store is configured to be
+    * {@link StoreConfiguration#async()}. This allows for this store to be unavailable for short periods of time
+    * until its buffer fills.
```

Comment:
"If a store is configured to be {@link StoreConfiguration#async()} and the store becomes unavailable, then it's possible for the cache operations to be accepted in the interim period between the loss of availability and the modification-queue becoming full. This allows for this store to be unavailable for short periods of time without a {@link StoreUnavailableException} being thrown, however if the store does not become available before the queue fills, then a {@link StoreUnavailableException} is eventually thrown.